# Stage 4: Image Preprocessing

## Objective

Prepare microbial images for deep learning model training.

Tasks:

1. Verify valid images
2. Skip corrupted images
3. Create labels 
4. label Encoding
5. Resize images
6. Normalize pixel values
7. Train-test split

### Task 1: Verify Valid Images

Total images
Valid images
Corrupted images

In [1]:
# ============================================================
# Stage 4: Verify Dataset Images
#
# Purpose:
# Count total images, valid images and corrupted images.
# ============================================================

# Import required libraries
import os
from PIL import Image

# Dataset path
dataset_path = "../archive"

# Store total image count
total_images = 0

# Store valid image count
valid_images = 0

# Store corrupted image count
corrupted_images = 0

# Traverse dataset folders
for species in os.listdir(dataset_path):

    # Create complete species path
    species_path = os.path.join(dataset_path, species)

    # Skip non-folder items
    if not os.path.isdir(species_path):
        continue

    # Traverse folders recursively
    for root, dirs, files in os.walk(species_path):

        # Check every file
        for file in files:

            # Process TIFF images only
            if file.lower().endswith(".tif"):

                # Increase total image count
                total_images += 1

                try:

                    # Create image path
                    image_path = os.path.join(root, file)

                    # Open image
                    image = Image.open(image_path)

                    # Verify image integrity
                    image.verify()

                    # Increase valid image count
                    valid_images += 1

                except Exception:

                    # Increase corrupted image count
                    corrupted_images += 1

# Print final report
print("DATASET VERIFICATION REPORT")
print()

print("Total Images     :", total_images)
print("Valid Images     :", valid_images)
print("Corrupted Images :", corrupted_images)

DATASET VERIFICATION REPORT

Total Images     : 689
Valid Images     : 689
Corrupted Images : 0


## Observation

Dataset verification was completed successfully.

Results:

- Total Images: 689
- Valid Images: 689
- Corrupted Images: 0

All images are valid and ready for preprocessing.

In [2]:
# ============================================================
# Stage 4: Create Image Paths and Labels
#
# Purpose:
# Collect image paths and corresponding species labels.
# ============================================================

# Import required libraries
import os

# Dataset path
dataset_path = "../archive"

# Store image paths
image_paths = []

# Store labels
labels = []

# Traverse species folders
for species in os.listdir(dataset_path):

    # Create complete species path
    species_path = os.path.join(dataset_path, species)

    # Skip non-folder items
    if not os.path.isdir(species_path):
        continue

    # Traverse folders recursively
    for root, dirs, files in os.walk(species_path):

        # Check every file
        for file in files:

            # Process TIFF images only
            if file.lower().endswith(".tif"):

                # Create complete image path
                image_path = os.path.join(root, file)

                # Store image path
                image_paths.append(image_path)

                # Store species label
                labels.append(species)

# Print dataset information
print("Total Image Paths :", len(image_paths))
print("Total Labels      :", len(labels))

# Display first 5 samples
for i in range(5):

    print()
    print("Image :", image_paths[i])
    print("Label :", labels[i])

Total Image Paths : 689
Total Labels      : 689

Image : ../archive/Proteus/Proteus_0008.tif
Label : Proteus

Image : ../archive/Proteus/Proteus_0020.tif
Label : Proteus

Image : ../archive/Proteus/Proteus_0009.tif
Label : Proteus

Image : ../archive/Proteus/Proteus_0019.tif
Label : Proteus

Image : ../archive/Proteus/Proteus_0018.tif
Label : Proteus


### Verify

In [3]:
# Print unique species count

print("Total Unique Species:", len(set(labels)))

print()

# Print all species names
for species in sorted(set(labels)):
    print(species)

Total Unique Species: 33

Acinetobacter.baumanii
Actinomyces.israeli
Bacteroides.fragilis
Bifidobacterium.spp
Candida.albicans
Clostridium.perfringens
Enterococcus.faecalis
Enterococcus.faecium
Escherichia.coli
Fusobacterium
Lactobacillus.casei
Lactobacillus.crispatus
Lactobacillus.delbrueckii
Lactobacillus.gasseri
Lactobacillus.jehnsenii
Lactobacillus.johnsonii
Lactobacillus.paracasei
Lactobacillus.plantarum
Lactobacillus.reuteri
Lactobacillus.rhamnosus
Lactobacillus.salivarius
Listeria.monocytogenes
Micrococcus.spp
Neisseria.gonorrhoeae
Porfyromonas.gingivalis
Propionibacterium.acnes
Proteus
Pseudomonas.aeruginosa
Staphylococcus.aureus
Staphylococcus.epidermidis
Staphylococcus.saprophiticus
Streptococcus.agalactiae
Veionella


In [5]:
# Show first 20 labels

for i in range(25):
    print(labels[i])

Proteus
Proteus
Proteus
Proteus
Proteus
Proteus
Proteus
Proteus
Proteus
Proteus
Proteus
Proteus
Proteus
Proteus
Proteus
Proteus
Proteus
Proteus
Proteus
Proteus
Lactobacillus.plantarum
Lactobacillus.plantarum
Lactobacillus.plantarum
Lactobacillus.plantarum
Lactobacillus.plantarum


In [6]:
# Total unique species

print(len(set(labels)))

33


### Label Encoding Kya Hota Hai?
Abhi labels text format me hain:
Proteus
Escherichia.coli
Staphylococcus.aureus
...
CNN text nahi samajhta.
Hume convert karna padega:


# Label Encoding

In [7]:
# ============================================================
# Stage 4: Label Encoding
#
# Purpose:
# Convert species names into numerical labels that can
# be used by deep learning models.
# ============================================================

# Import LabelEncoder
from sklearn.preprocessing import LabelEncoder

# Create LabelEncoder object
label_encoder = LabelEncoder()

# Convert text labels into numbers
encoded_labels = label_encoder.fit_transform(labels)

# Print total classes
print("Total Classes :", len(label_encoder.classes_))

# Display first 10 label mappings
print("\nFirst 10 Encoded Labels:\n")

for i in range(10):

    print(
        labels[i],
        " ---> ",
        encoded_labels[i]
    )

ModuleNotFoundError: No module named 'sklearn'

In [8]:
import sys

print(sys.executable)

/Users/viveksingh/AMR_PROJECT/venv/bin/python
